src/Microsoft.AspNetCore.Blazor.Razor.Extensions/ScopeStack.cs

```diff
@@ -26,19 +26,39 @@ public void OpenElementScope(string tagName)
             _stack.Push(new ScopeEntry(tagName, ScopeKind.Element));
         }
 
-        public void OpenComponentScope(string tagName)
+        public void OpenComponentScope(CodeRenderingContext context, string name, string type, string parameterName)
         {
-            _stack.Push(new ScopeEntry(tagName, ScopeKind.Component));
+            var scope = new ScopeEntry(name, ScopeKind.Component);
+            _stack.Push(scope);
+
+            var blazorNodeWriter = (BlazorNodeWriter)context.NodeWriter;
+            blazorNodeWriter.BeginWriteAttribute(context.CodeWriter, name);
+            OffsetBuilderVarNumber(1);
+
+            // Writes code that looks like:
+            //
+            // builder.AddAttribute(0, "{name}", ({type})((__builder) => { ... }));
+            // OR
+            // builder.AddAttribute(0, "{name}", ({type})((context) => (__builder) => { ... }));
+
+            context.CodeWriter.Write($"({type})(");
+
+            if (parameterName != null)
+            {
+                context.CodeWriter.Write($"({parameterName}) => ");
+            }
+
+            scope.LambdaScope = context.CodeWriter.BuildLambda(BuilderVarName);
```

Comment:
Previously, the way we avoided emitting unnecessary empty lambdas was only to write them when we observed the first child item within a component.

If I'm following the new implementation correctly, it now works by creating a ComponentChildContent node, but only when there's some nonempty content inside the component. And then it only creates the lambda when writing the ComponentChildContent node, hence no further need to call `IncrementCurrentScopeChildCount`.

Have I followed that correctly? If so, looks good to me!

Reply:
yes that's right. We need to know while we're building the tree if you have non-trivial content inside a component, so that knowledge is now represented in the tree instead of in the writer.